Structure lifecycle hooks, add links to build time hooks

lukastaegert opened the pull request:
<!--
  ⚡️ katchow! We ❤️ Pull Requests!

  If you remove or skip this template, you'll make the 🐼 sad and the mighty god
  of Github will appear and pile-drive the close button from a great height
  while making animal noises.

  Pull Request Requirements:
  * Please include tests to illustrate the problem this PR resolves.
  * Please lint your changes by running `npm run lint` before creating a PR.
  * Please update the documentation in `/docs` where necessary

  Please place an x (no spaces - [x]) in all [ ] that apply.
-->

This PR contains:
- [ ] bugfix
- [ ] feature
- [ ] refactor
- [x] documentation
- [ ] other

Are tests included?
- [ ] yes (*bugfixes and features will not be merged without tests*)
- [x] no

Breaking Changes?
- [ ] yes (*breaking changes will not be merged unless absolutely necessary*)
- [x] no

List any relevant issue numbers:
Resolves #3409

### Description
This will
- split up the hooks in the documentation into build phase and output generation phase hooks
- mention in the beginnings of these sections which are the first and last hooks of each phase
- add to each hook which hooks precede and follow it

This is still WIP. The build phase is already done, I hope to finish the output generation phase until tomorrow. Feedback already welcome.

jakearchibald commented on docs/05-plugin-development.md:
```diff
@@ -63,16 +63,177 @@ Type: `string`
+#### `watchChange`
+Type: `(id: string) => void`<br>
+Kind: `sync, sequential`<br>
+Previous/Next Hook: This hook can be triggered at any time both during the build and the output generation phases.
```
What happens once a file has changed? Does it go back to the start, as in `options`?

lukastaegert replied:
I will add some details

jakearchibald commented on docs/05-plugin-development.md:
```diff
@@ -63,16 +63,177 @@ Type: `string`
+#### `transform`
+Type: `(code: string, id: string) => string | null | { code: string, map?: string | SourceMap, ast? : ESTree.Program, moduleSideEffects?: boolean | null, syntheticNamedExports?: boolean | null }`<br>
+Kind: `async, sequential`<br>
+Previous Hook: [`load`](guide/en/#load) where the transformed file was loaded.<br>
```
It seems a little confusing that the file is already a 'transformed file'

lukastaegert commented on docs/05-plugin-development.md:
```diff
@@ -63,16 +63,177 @@ Type: `string`
+### Build Hooks
 
-In addition to properties defining the identity of your plugin, you may also specify properties that correspond to available build hooks. Hooks can affect how a build is run, provide information about a build, or modify a build once complete. There are different kinds of hooks:
+To make your plugin functional, you add function properties that correspond to some of the various available hooks. They will be called by Rollup at specific points in the build process. Hooks can affect how a build is run, provide information about a build, or modify a build once complete. There are different kinds of hooks:
```
Nice one, thanks!

lukastaegert commented on docs/05-plugin-development.md:
```diff
@@ -63,16 +63,177 @@ Type: `string`
+* `parallel`: If several plugins implement this hook, all of them will be run in the specified plugin order. If a hook is async, subsequent hooks of this kind will be run in parallel and not wait for the current hook.
 
-Furthermore, hooks can be run either during the `build` phase of the Rollup build, which is triggered by `rollup.rollup()`, or during the `generate` phase, which is triggered by `bundle.generate()`  or `bundle.write()`. Plugins that only use `generate` phase hooks can also be passed in via the output options to `bundle.generate()` or `bundle.write()` and therefore run only for certain outputs.
+Build hooks are run during the build phase, which is triggered by `rollup.rollup(inputOptions)`. They are mainly concerned with locating, providing and transforming input files before they are processed by Rollup. The first hook of the build phase is [options](guide/en/#options), the last one is always [buildEnd](guide/en/#buidend). Additionally, the [watchChange](guide/en/#watchchange) hook can  ...
```
Yes, at the moment there is no mechanism to prevent this from happening. A rebuild, however, will only start once the current build is complete and sufficient time, at the moment 200ms, has passed without another change event. Maybe I should spell this out in more detail.

lukastaegert commented:
I finished adding the remaining links to the output generation hooks and addressing the current review comments. Please have a look if you see anything that could be improved. Also note that I am still no native speaker so if you think something could be spelled out in a more natural way, do not hesitate to comment 😜

lukastaegert commented:
> This is looking great. The split between 'build' and 'output generation' really improved my understanding here.

Thanks. And yes, this is something that I stumbled upon myself while extending the documentation. E.g. resolveId and resolveDynamicImport are a build phase hook while resolveFileUrl and resolveImportMeta are output generation hooks, the naming does not really help here. As most plugins will probably be concerned mainly with one of the two phases, this split seemed helpful to me and you confirmed this. In retrospective, `renderFileUrl` and `renderImportMeta` would have probably been better names 🙄

jakearchibald commented on docs/05-plugin-development.md:
```diff
@@ -63,21 +63,183 @@ Type: `string`
+* `parallel`: If several plugins implement this hook, all of them will be run in the specified plugin order. If a hook is async, subsequent hooks of this kind will be run in parallel and not wait for the current hook.
 
-Furthermore, hooks can be run either during the `build` phase of the Rollup build, which is triggered by `rollup.rollup()`, or during the `generate` phase, which is triggered by `bundle.generate()`  or `bundle.write()`. Plugins that only use `generate` phase hooks can also be passed in via the output options to `bundle.generate()` or `bundle.write()` and therefore run only for certain outputs.
+Build hooks are run during the build phase, which is triggered by `rollup.rollup(inputOptions)`. They are mainly concerned with locating, providing and transforming input files before they are processed by Rollup. The first hook of the build phase is [options](guide/en/#options), the last one is always [buildEnd](guide/en/#buidend). Additionally, the [watchChange](guide/en/#watchchange) hook can  ...
```
This clarification to `watchChange` is great!

jakearchibald replied:
Potential gotcha: When you say 'build' here, does that mean the steps contained in 'build hooks' or does that also include 'output generation'?

lukastaegert replied:
In the end it means both, I will replace it with "run"

jakearchibald commented on docs/05-plugin-development.md:
```diff
@@ -225,20 +225,21 @@ You can use [`this.getModuleInfo`](guide/en/#thisgetmoduleinfomoduleid-string--m
 Type: `(id: string) => void`<br>
 Kind: `sync, sequential`<br>
-Previous/Next Hook: This hook can be triggered at any time both during the build and the output generation phases.
+Previous/Next Hook: This hook can be triggered at any time both during the build and the output generation phases. If that is the case, the current build will still proceed but a new build will be scheduled to start once the current build has completed, starting again with [`options`](guide/en/#options).
```
This is great! I was never sure if build re-runs were sometimes shortcut. This makes it clear.

jakearchibald commented on docs/05-plugin-development.md:
```diff
@@ -257,23 +258,23 @@ augmentChunkHash(chunkInfo) {
 Kind: `async, parallel`<br>
 Previous Hook: [`renderStart`](guide/en/#renderstart)<br>
-Next Hook: 
+Next Hook: [`augmentChunkHash`](guide/en/#augmentchunkhash) for each chunk with that would contain a hash in the file name. Otherwise [`resolveFileUrl`](guide/en/#resolvefileurl) for each use of `import.meta.ROLLUP_FILE_URL_referenceId` and [`resolveImportMeta`](guide/en/#resolveimportmeta) for all other accesses to `import.meta`. Otherwise [`renderChunk`](guide/en/#renderchunk) for each chunk.
```
The use of "otherwise" tripped me up here. It makes it sound like `resolveFileUrl` won't happen if `augmentChunkHash` happens, but you mean "otherwise `resolveFileUrl` will happen next". Maybe I just wasn't reading closely and I'll be the only one who gets confused here.

"Otherwise" could probably be replaced with "then", or just remove those statements and only mention `resolveFileUrl` as the "next hook" to `augmentChunkHash`.

lukastaegert replied:
I will use `then`

jakearchibald commented:
In case it's useful, here's how I interpreted the steps:

To **build**:

1. Hook: `options`.
1. Hook: `buildStart`.
1. For each entry point, in parallel, go to step 2 of **load module**, and wait for all to complete.
1. Hook: `buildEnd`.
1. If previous steps did not error, for each set of output options, go to step 1 of **generate output**.
1. If in watch mode:
    1. If a watched file has changed, go to 1.
    1. Wait for a watched file to change.
    1. Go to 1. 

To **load module**:

1. Hook: `resolveDynamicImport` for this entry point's ID.
1. If not resolved, hook: `resolveId` for this entry point's ID.
1. Hook: `load` using the resolved ID.
1. Hook: `transform` using the loaded module.
1. Perform these in parallel, and wait for all to complete:
    - For each import within the transformed module, in parallel, go to step 2 of **load module**, and wait for all to complete.
    - For each dynamic import within the transformed module, in parallel, go to step 1 of **load module**, and wait for all to complete.

To **generate output**:

1. Hook: `outputOptions`.
1. Try:
    1. Hook: `renderStart`.
    1. Perform these in parallel, and wait for all to complete:
        - Hook: `banner`
        - Hook: `footer`
        - Hook: `intro`
        - Hook: `outro`
    1. For each chunk that will contain a hash in its output filename, hook: `augmentChunkHash`.
    1. For each chunk, for each use of `import.meta.ROLLUP_FILE_URL_referenceId`, hook: `resolveFileUrl`.
    1. For each chunk, for each use of `import.meta`, hook: `resolveImportMeta`.
    1. For each chunk, hook: `renderChunk`.
1. Catch:
    1. Hook: `renderError`.
    1. Return.
1. Hook: `generateBundle`.
1. If writing output, hook: `writeBundle`.

If any watched file changes, immediately hook: `watchChange`.

lukastaegert commented:
Nice, this looks quite good! Minor points:
* While watch mode is correct in essence, the way you wrote it it does not belong to the build phase but rather to the generate phase. Watch mode KNOWS how many outputs there are. The logic is: Once the last output has been generated and a file has change, go to build 1 (otherwise wait etc.)
* `resolveFileUrl` and `resolveImportMeta` are not two separate loops but rather the same loop that just looks for all usages of `import.meta.something` and then executes either of the two depending on `something`
* `generateBundle` also has a possible `return` depending on how Rollup was called

lukastaegert commented:
> Looks good, although it still isn't clear to be when Rollup performs hashing, dead code removal, splitting etc.

Well, that was not the question you asked 😉. So to answer it: hashing is performed just after the `augmentChunkHash` is called (basically it is called as part of the process). But I would not want to document this as this is definitely about to change once the hashing is redone.

Dead code elimination (or rather "tree-shaking" as we like to call it here 😄) is performed once all modules are loaded, so between the last `transform` and before `buildEnd`. This is something that we could document as I expect this to remain stable.

Code-splitting is also done in the build phase at the moment, after tree-shaking and before `buildEnd`. This is something I would not want to document as I very much want to move this to the output generation phase at some point. It is not too costly but having code-splitting per output would allow to make `inlineDynamicImports` an output option or have output-specific helpers injected as virtual modules into the build, e.g. helpers to unpack external namespaces from AMD or CJS.

jakearchibald commented:
> The logic is: Once the last output has been generated and a file has change, go to build 1 (otherwise wait etc.)

Isn't that what I wrote? See step 6 of **build**.

> * `resolveFileUrl` and `resolveImportMeta` are not two separate loops but rather the same loop that just looks for all usages of `import.meta.something` and then executes either of the two depending on `something`

Ohh nice! I've updated the post. What about `renderChunk`, is that a separate loop?

> * `generateBundle` also has a possible `return` depending on how Rollup was called

Isn't that covered in the next step: If writing…

> > Looks good, although it still isn't clear to be when Rollup performs hashing, dead code removal, splitting etc.
> 
> Well, that was not the question you asked 😉.

Hmmm kinda? In https://github.com/rollup/rollup/issues/3409#issue-572626720 I call out that it's currently difficult to figure out the order in which hooks are called, and *how it relates to the rest of Rollup's processing*. I also called out https://github.com/rollup/rollup/issues/2739 - it isn't clear from the docs that `renderChunk` comes after hashing.

> So to answer it: hashing is performed just after the `augmentChunkHash` is called (basically it is called as part of the process)

Is that also when asset hashing happens?

> But I would not want to document this as this is definitely about to change once the hashing is redone.

Hmm ok. Is hashing likely to be redone in the next 6 months? Otherwise it might be worth noting it, to avoid more confusion like https://github.com/rollup/rollup/issues/2739.

I've added the other steps too, but noted that hashing and splitting are likely to move. Thanks for taking the time to work on this!

lukastaegert commented:
> Isn't that what I wrote? See step 6 of build.
Yes, but it does not happen at the end of "build", it happens at the end of "generate output" for the last output

> Ohh nice! I've updated the post. What about renderChunk, is that a separate loop?
Yes

 > Isn't that covered in the next step: If writing…
Ah, I overlooked this 👍

> Hmmm kinda? In #3409 (comment) I call out that it's currently difficult to figure out the order in which hooks are called, and how it relates to the rest of Rollup's processing. I also called out #2739 - it isn't clear from the docs that renderChunk comes after hashing

Ok sorry, seems my mind had simplified this a little too much.

> Is that also when asset hashing happens?

No, those are currently hashed when they are emitted or their source is set as you cannot change the source later.

> Hmm ok. Is hashing likely to be redone in the next 6 months

I would think so. Documentation alone will not prevent new bugs from appearing here I fear.

> I've added the other steps too, but noted that hashing and splitting are likely to move. Thanks for taking the time to work on this!

And thanks for pushing here! There are some comments from you that I will also add later before publishing this.
